File: DSLKIT/DLSKit/Core/DSLModifierRegistry.swift
//
//  DSLModifierRegistry.swift
//  HelloWorld
//
//  Created by Pixel Logic Apps on 25/04/25.
//

import SwiftUI

public class DSLModifierRegistry<ViewType> {
    public typealias Modifier = (ViewType, Any, DSLContext) -> ViewType

    private var modifiers: [String: Modifier] = [:]

    public func register(_ name: String, _ fn: @escaping Modifier) {
        modifiers[name] = fn
    }

    public func apply(_ list: [[String: Any]], to view: ViewType, context: DSLContext) -> ViewType {
        logDebug("--- DEBUG: Modifier Apply - STARTING application for list: \(list)")
        return list.reduce(view) { current, mod in
            logDebug("--- DEBUG: Modifier Apply - Processing mod dict: \(mod)")
            guard let key = mod.keys.first,
                  let value = mod[key],
                  let fn = modifiers[key] else {
                // Log se o modificador for pulado
                logDebug("--- DEBUG: Modifier Apply - SKIPPING modifier (key not found or fn not registered): \(mod)")
                return current
            }
            // Log se o modificador for aplicado
            logDebug("--- DEBUG: Modifier Apply - APPLYING modifier: \(key)")
            return fn(current, value, context)
        }
    }
}

File: compose-dsl/src/main/java/com/example/ComposeDSL/DSLModifierRegistry.kt
package com.example.ComposeDSL

import androidx.compose.ui.Modifier

/** Simple registry to apply modifiers from JSON nodes */
object DSLModifierRegistry {
    private val registry = mutableMapOf<String, (Modifier, Any?, DSLContext) -> Modifier>()

    fun register(name: String, fn: (Modifier, Any?, DSLContext) -> Modifier) {
        registry[name] = fn
    }

    fun apply(list: List<Map<String, Any?>>, base: Modifier, context: DSLContext): Modifier {
        var result = base
        for (mod in list) {
            val key = mod.keys.firstOrNull() ?: continue
            val value = mod[key]
            val fn = registry[key] ?: continue
            result = fn(result, value, context)
        }
        return result
    }

    fun registerDefaults() {
        BaseViewModifiers.registerAll()
    }
}
